content/issue-1/2023-kcl-rewrite-with-rust.md

```diff
@@ -0,0 +1,139 @@
+
+## Mental transformation
+
+First of all, the syntax and semantics of Rust well absorb and integrate the concepts related to the type system in functional programming, such as the Abstract Algebraic Type (ADT). In addition, there is no concept related to "inheritance" in Rust. If you can't understand it well, even ordinary structure definitions in other languages may take a lot of time in Rust. For example, the following Python code may be defined like this in Rust
+
+- Python
+
+```python
+from dataclasses import dataclass
+
+class KCLObject:
+    pass
+
+@dataclass
+class KCLIntObject(KCLObject):
+    value: int
+
+@dataclass
+class KCLFloatObject(KCLObject):
+    value: float
+```
+
+- Rust
+
+```rust
+enum KCLObject {
+    Int(u64),
+    Float(f64),
+}
+```
+
+Of course, more time is spent fighting against the error reports of the Rust compiler itself. The Rust compiler will often cause developers to "run into a wall", such as borrowing check errors. Especially for the KCL compiler, its core structure is the Abstract Syntax Tree (AST), which is a recursive and nested tree structure.
+
+It is sometimes difficult to give consideration to the relationship between variable variability and borrowing check in Rust, Just like the scope structure `Scope` defined in KCL compiler, for scenarios with circular references, it is used to display the interdependence of data that needs to be aware of, while making extensive use of intelligent pointer structures commonly used in Rust such as `Rc`, `RefCell` and `Weak`.
+
+```rust
+/// A Scope maintains a set of objects and links to its containing
+/// (parent) and contained (children) scopes. Objects may be inserted
+/// and looked up by name. The zero value for Scope is a ready-to-use
+/// empty scope.
+#[derive(Clone, Debug)]
+pub struct Scope {
+    /// The parent scope.
+    pub parent: Option<Weak<RefCell<Scope>>>,
+    /// The child scope list.
+    pub children: Vec<Rc<RefCell<Scope>>>,
+    /// The scope object mapping with its name.
+    pub elems: IndexMap<String, Rc<RefCell<ScopeObject>>>,
+    /// The scope start position.
+    pub start: Position,
+    /// The scope end position.
+    pub end: Position,
+    /// The scope kind.
+    pub kind: ScopeKind,
+}
+```
+
+## Development efficiency
+
+The development efficiency of Rust can be described as "restraining first and then improving". At the beginning of the handwritten project, if the team members have not been exposed to the concept of functional programming and related programming habits, the development speed will be significantly slower than Python, Go, Java and other languages. However, once they become familiar with the common methods and best practices of the Rust standard library, as well as the common error modification of the Rust compiler, the development efficiency will be greatly improved, and they can write high-quality, safe and efficient code natively.
```

Comment:
I don't understand what that phrase means here, nor "the handwritten project" to be honest

Reply:
> I don't understand what that phrase means here, nor "the handwritten project" to be honest

fixed.

Reply:
> It would be great to hear more about the 40 X - how did you benchmark it at 40 X and what language was it in before? Did it take a long time to rewrite it? Were there some particular steps that were bottleneck processes like maybe some iteration process that was taking a long time to execute in the the first language? Why did you choose Rust?

Thank you for your feedback. In the article, I explained that before using Rust, we used Python to write the entire project. Then we do a benchmark on virtual machine programs for Rust, Go, and Python, and compared the time cost. Finally, after comprehensive consideration, we chose Rust and spent several months rewriting the entire project. In addition, due to article length constraints, more content will be discussed in later articles.